Adopt dynamicDowncast<> in HitTestResult and RenderBoxModelObject

annevk opened the pull request:
#### aafdc620e2ab058a8e2c106e1062362afe128f69
<pre>
Adopt dynamicDowncast&lt;&gt; in HitTestResult and RenderBoxModelObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=267220">https://bugs.webkit.org/show_bug.cgi?id=267220</a>

Reviewed by Darin Adler.

All downcasts are covered in HitTestResult. RenderBoxModelObject has
some remaining work.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::setInnerNode):
(WebCore::HitTestResult::setInnerNonSharedNode):
(WebCore::HitTestResult::title const):
(WebCore::HitTestResult::innerTextIfTruncated const):
(WebCore::HitTestResult::altDisplayString const):
(WebCore::HitTestResult::image const):
(WebCore::HitTestResult::absoluteImageURL const):
(WebCore::HitTestResult::absolutePDFURL const):
(WebCore::HitTestResult::mediaElement const):
(WebCore::HitTestResult::enterFullscreenForVideo const):
(WebCore::HitTestResult::isOverTextInsideFormControlElement const):
(WebCore::HitTestResult::isContentEditable const):
(WebCore::HitTestResult::targetElement const):
(WebCore::HitTestResult::innerNonSharedElement const):
(WebCore::HitTestResult::toggleEnhancedFullscreenForVideo const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::hasAutoHeightOrContainingBlockWithAutoHeight const):
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
(WebCore::RenderBoxModelObject::adjustedPositionRelativeToOffsetParent const):
(WebCore::RenderBoxModelObject::paintMaskForTextFillBox):
(WebCore::RenderBoxModelObject::inlineContinuation const):
(WebCore::RenderBoxModelObject::collectAbsoluteQuadsForContinuation const):

Canonical link: <a href="https://commits.webkit.org/272791@main">https://commits.webkit.org/272791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ff77c68ba1edac9fe1a343b2e1a51550b4c8318

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34929 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33996 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8968 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29270 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9948 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8668 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29466 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37001 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34952 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8960 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32807 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10652 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7666 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9551 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9593 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -809,19 +795,19 @@ Vector<String> HitTestResult::dictationAlternatives() const
+    auto* node = m_innerNonSharedNode.get();
     if (!node)
         return nullptr;
-    if (is<Element>(*node))
-        return downcast<Element>(node);
+    if (auto* element = dynamicDowncast<Element>(*node))
+        return element;
     return node->parentElement();
```
Seems like this operation "self if element, parent element otherwise" should be a named function.

darinadler commented on Source/WebCore/rendering/HitTestResult.cpp:
```diff
@@ -413,7 +405,7 @@ URL HitTestResult::absoluteImageURL() const
         || is<HTMLObjectElement>(*imageNode)
         || is<SVGImageElement>(*imageNode)) {
-        auto imageURL = imageNode->document().completeURL(downcast<Element>(*imageNode).imageSourceURL());
+        auto imageURL = imageNode->document().completeURL(uncheckedDowncast<Element>(*imageNode).imageSourceURL());
```
Seems better to do a `dynamicDowncast<Element>` first and do all the `is<>` on that, and then not have an `uncheckedDowncast` here.

annevk committed:
Adopt dynamicDowncast<> in HitTestResult and RenderBoxModelObject
https://bugs.webkit.org/show_bug.cgi?id=267220

Reviewed by Darin Adler.

All downcasts are covered in HitTestResult. RenderBoxModelObject has
some remaining work.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::setInnerNode):
(WebCore::HitTestResult::setInnerNonSharedNode):
(WebCore::HitTestResult::title const):
(WebCore::HitTestResult::innerTextIfTruncated const):
(WebCore::HitTestResult::altDisplayString const):
(WebCore::HitTestResult::image const):
(WebCore::HitTestResult::absoluteImageURL const):
(WebCore::HitTestResult::absolutePDFURL const):
(WebCore::HitTestResult::mediaElement const):
(WebCore::HitTestResult::enterFullscreenForVideo const):
(WebCore::HitTestResult::isOverTextInsideFormControlElement const):
(WebCore::HitTestResult::isContentEditable const):
(WebCore::HitTestResult::targetElement const):
(WebCore::HitTestResult::innerNonSharedElement const):
(WebCore::HitTestResult::toggleEnhancedFullscreenForVideo const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::hasAutoHeightOrContainingBlockWithAutoHeight const):
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
(WebCore::RenderBoxModelObject::adjustedPositionRelativeToOffsetParent const):
(WebCore::RenderBoxModelObject::paintMaskForTextFillBox):
(WebCore::RenderBoxModelObject::inlineContinuation const):
(WebCore::RenderBoxModelObject::collectAbsoluteQuadsForContinuation const):

Canonical link: https://commits.webkit.org/272791@main
